pkg/eventmonitor/consumers/process.go

```diff
@@ -0,0 +1,171 @@
+func NewProcessConsumer(id string, chanSize int, evm *eventmonitor.EventMonitor) (*ProcessConsumer, error) {
+	pc := &ProcessConsumer{
+		id:            id,
+		chanSize:      chanSize,
+		execCallbacks: callbackMap{callbacks: make(map[*ProcessCallback]struct{})},
+		exitCallbacks: callbackMap{callbacks: make(map[*ProcessCallback]struct{})},
+	}
+
+	if err := evm.AddEventConsumerHandler(pc); err != nil {
+		return nil, fmt.Errorf("cannot add event consumer handler: %w", err)
+	}
+
+	evm.RegisterEventConsumer(pc)
+
+	return pc, nil
+}
+
+// --- eventmonitor.EventConsumer interface methods
+
+// Start starts the consumer, this is a no-op for this implementation
+func (p *ProcessConsumer) Start() error {
+	return nil
+}
+
+// Stop stops the consumer, this is a no-op for this implementation
+func (p *ProcessConsumer) Stop() {
+}
+
+// ID returns the ID of the consumer
+func (p *ProcessConsumer) ID() string {
+	return p.id
+}
+
+// --- eventmonitor.EventConsumerHandler interface methods
+
+// ChanSize returns the size of the channel that the event monitor will use to send events to this handler
+func (p *ProcessConsumer) ChanSize() int {
+	return p.chanSize
+}
+
+// EventTypes returns the event types that this handler is interested in
+func (p *ProcessConsumer) EventTypes() []model.EventType {
+	return []model.EventType{model.ExecEventType, model.ExitEventType}
+}
+
+// Copy copies the event from the eventmonitor type to the one we use in this struct
+func (p *ProcessConsumer) Copy(ev *model.Event) any {
+	return &event{
+		eventType: ev.GetEventType(),
+		pid:       ev.GetProcessPid(),
+	}
+}
+
+// HandleEvent handles the event from the event monitor
+func (p *ProcessConsumer) HandleEvent(ev any) {
+	sevent, ok := ev.(*event)
+	if !ok {
+		return
+	}
+
+	switch sevent.eventType {
+	case model.ExecEventType:
+		p.execCallbacks.call(sevent.pid)
+	case model.ExitEventType:
+		p.exitCallbacks.call(sevent.pid)
+	}
+}
+
+// SubscribeExec subscribes to process exec events, and returns the function
+// that needs to be called to unsubscribe
+func (p *ProcessConsumer) SubscribeExec(callback ProcessCallback) func() {
+	return p.execCallbacks.add(callback)
+}
+
+// SubscribeExit subscribes to process exit events, and returns the function
+// that needs to be called to unsubscribe
+func (p *ProcessConsumer) SubscribeExit(callback ProcessCallback) func() {
+	return p.exitCallbacks.add(callback)
+}
+
+// callbackMap is a helper struct that holds a map of callbacks and a mutex to protect it
+type callbackMap struct {
+	// callbacks holds the set of callbacks
+	callbacks map[*ProcessCallback]struct{}
+
+	// mutex is the mutex that protects the callbacks map
+	mutex sync.RWMutex
+
+	// hasCallbacks is a flag that indicates if there are any callbacks subscribed, used
+	// to avoid locking/unlocking the mutex if there are no callbacks
+	hasCallbacks atomic.Bool
```

Comment:
imho this is unnecessary complexity.
there are some cases that you would miss, since it is not guarantee a fully safe concurrent operations (this [line](https://github.com/DataDog/datadog-agent/pull/30559/files#diff-42bf6324ee3c6c4e49f16275eb9ab8326a280d0abc72b8601d8393d62c7505f6R157) is not an atomic operation)

consider removing this field for simplicity, see my other comment about choice of the locking mechanism. if you decide to perform optimizations in the first place, i suggest performing benchmarks on the rwlock and see if that is actually a good implementation choice

Reply:
That code comes from the process-monitor existing implementation. I found the PR that introduced the analogous atomic in that code https://github.com/DataDog/datadog-agent/pull/17457, it seems this avoids extra CPU usage during startup when there are no subscribers, and it doesn't add runtime cost when there are subscribers.